### ISSUE TITLE
Gradient of `jnp.linalg.norm` does not respect absolute homogeneity.

### ISSUE THREAD

**Issue by tillahoffmann** [jax-ml/jax#26248]
### Description

The [norm](https://en.wikipedia.org/wiki/Norm_(mathematics)) should satisfy $d= \left\Vert a \mathbf{b}\right\Vert = \left\vert a\right\vert \left\Vert\mathbf{b}\right\Vert$ for scalar $a$ and vector $\mathbf{b}$. Consequently $\frac{\partial d}{\partial a} = \mathrm{sign}\left(a\right) \left\Vert\mathbf{b}\right\Vert$. However, the gradient of the standard implementation of the $p$-norm $\left(\sum_{i} b_i ^ p\right)^{1/p}$ is not defined at $\mathbf{b}=\mathbf{0}$ because applying the chain rule includes terms involving negative powers of zero.

```python
>>> import jax
>>> from jax import numpy as jnp

>>> def func1(a, x):
...     return jnp.linalg.norm(a * x)


>>> def func2(a, x):
...     return jnp.abs(a) * jnp.linalg.norm(x)


>>> funcs = [func1, func2]
>>> a = 1.3
>>> x = jnp.zeros(2)

>>> for func in funcs:
...     print(f"{func.__name__}({a}, {x}) = {func(a, x)}")
...     print(f"jax.grad({func.__name__})({a}, {x}) = {jax.grad(func)(a, x)}")

func1(1.3, [0. 0.]) = 0.0
jax.grad(func1)(1.3, [0. 0.]) = nan
func2(1.3, [0. 0.]) = 0.0
jax.grad(func2)(1.3, [0. 0.]) = 0.0
```

This particular situation arises, for example, in the evaluation of the diagonal of the [Matern covariance kernel](https://en.wikipedia.org/wiki/Matérn_covariance_function) for Gaussian processes in more than one dimension.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.0
jaxlib: 0.5.0
numpy:  2.2.1
python: 3.11.5 (main, Dec  8 2023, 17:04:09) [Clang 15.0.0 (clang-1500.0.40.1)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='Tills-MacBook-Pro-3.local', release='24.2.0', version='Darwin Kernel Version 24.2.0: Fri Dec  6 18:40:14 PST 2024; root:xnu-11215.61.5~2/RELEASE_ARM64_T8103', machine='arm64')
```

**Comment by jakevdp** [comment]
Ping @mattjj because he daydreams about autodiff corner cases

**Comment by mattjj** [comment]
Sorry for being slow to respond. Thanks for the clear exposition!

@tillahoffmann can you be more specific about the request here?

IIUC, you are dissatisfied that writing the same mathematical function two different ways as Python functions results in different autodiff derivatives. After all, there's one mathematical function, and one derivative function, so why should the results depend on the particular spelling as a Python function?

But autodiff does depend on the spelling, and in more cases than just this one. Autodiff is fundamentally a transformation on programs, and we require conditions on the particular programs (i.e. spellings), not just on the mathematical functions they denote, for it to succeed.

We can conjure lots of examples of Python functions with the same input-output behavior (i.e. denoting the same mathematical function) but that result in different derivatives. See #1052 for a few instances of one kind of example. But here are simpler ones:

```python
import jax
import jax.numpy as jnp

max1 = lambda x, y: x if x > y else y
max2 = jnp.maximum
print(jax.grad(max1)(1., 1.))  # 0.0
print(jax.grad(max2)(1., 1.))  # 0.5

f1 = lambda x: jnp.sqrt(x ** 2)
f2 = lambda x: jnp.abs(x)
print(jax.grad(f1)(0.))  # nan
print(jax.grad(f2)(0.))  # 1.0

g1 = lambda x, y: jnp.sqrt((x * y) ** 2)
g2 = lambda x, y: jnp.abs(x) * jnp.abs(y)
print(jax.grad(g1)(1.3, 0.))  # nan
print(jax.grad(g2)(1.3, 0.))  # 0.0

id1 = lambda x: x
id2 = lambda x: -x if x == 0 else x
print(jax.grad(id1)(0.))  #  1.0
print(jax.grad(id2)(0.))  # -1.0  !?!?!
```

I think the middle two examples are roughly analogous to our current situation here.

Having the autodiff result depend on the Python program spelling is unavoidable in general.

So is the concrete request to change something about the derivative for `jnp.linalg.norm` at zero? (Which you can always work around by using custom_jvp/custom_vjp, if you want.) Or something else?

**Comment by mattjj** [comment]
Maybe you want to write a function like

```python
@jax.custom_jvp  # or custom_vjp
def scaled_norm(a, x):
  return jnp.linalg.norm(a * x)

...
```

where you implement the derivative rule for this composition and thus get exactly the behavior you want.

**Comment by carlosgmartin** [comment]
@mattjj Are we open to defining the gradient of [norm](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.linalg.norm.html)/[vector_norm](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.linalg.vector_norm.html) at zero to be the zero vector? As with [abs](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.abs.html), the zero vector is in the convex hull of subgradients, and is the natural symmetric choice.

[This](https://github.com/jax-ml/jax/pull/27935) attempt was closed for inactivity.

It would be nice to at least have an argument flag to enable this behavior.

**Comment by mattjj** [comment]
I'm going to close this issue because I think it was solved.

@carlosgmartin as to your question, I think that attempt belonged to @rdyro. He tells me the reason it wasn't merged is it was hard to get correct second-order behavior depending on how different autodiff modes were mixed.

In general I'd say
1. I'm not against defining the gradient at zero to be zero, but there are some follow-up conventions to decide like defining the Hessian at zero (and similar for other higher-order derivatives);
2. in general it's not a super high priority so we're not actively working on it, though we'd welcome contributions!

**Comment by tillahoffmann** [comment]
Thank you for looking into this!

**Comment by amacati** [comment]
I came across this thread after facing the same `jnp.linalg.norm` NaN gradient issues described here and in several older threads. Just to double-check @mattjj: You would be open to change the behavior if someone worked on this and the second derivative issues would be sorted out?

**Comment by amacati** [comment]
> In general I'd say
>   1. I'm not against defining the gradient at zero to be zero, but there are some follow-up conventions to decide like defining the Hessian at zero (and similar for other higher-order derivatives);
>   2. in general it's not a super high priority so we're not actively working on it, though we'd welcome contributions!

@mattjj @jakevdp I recently stumbled across this issue again while optimizing functions that make heavy use of the norm where this is a common footgun. You previously mentioned you'd be open to define the gradient as 0 if a consistent rule for higher-order derivatives was found. Since the Hessian for the norm blows up around 0, I think it would be sensible to define any higher-order derivative as NaN. Would you still be open to a contribution that tackles this?
